gst/join/gstjoin.c

```diff
@@ -0,0 +1,1075 @@
+}
+
+/**
+ * @brief check if the pad is the active sinkpad.
+ */
+static inline gboolean
+gst_join_is_active_sinkpad (GstJoin * sel, GstPad * pad)
+{
+  gboolean res;
+
+  GST_JOIN_LOCK (sel);
+  res = (pad == sel->active_sinkpad);
+  GST_JOIN_UNLOCK (sel);
+
+  return res;
+}
+
+/**
+ * @brief Get or create the active sinkpad.
+ * @note  must be called with JOIN_LOCK.
+ */
+static GstPad *
+gst_join_get_active_sinkpad (GstJoin * sel)
+{
+  GstPad *active_sinkpad;
+
+  active_sinkpad = sel->active_sinkpad;
+  if (active_sinkpad == NULL) {
+    GValue item = G_VALUE_INIT;
+    GstIterator *iter = gst_element_iterate_sink_pads (GST_ELEMENT_CAST (sel));
+    GstIteratorResult ires;
+
+    while ((ires = gst_iterator_next (iter, &item)) == GST_ITERATOR_RESYNC)
+      gst_iterator_resync (iter);
+    if (ires == GST_ITERATOR_OK) {
+      /**
+       * If no pad is currently selected, we return the first usable pad to
+       * guarantee consistency
+       */
+
+      active_sinkpad = sel->active_sinkpad = g_value_dup_object (&item);
+      g_value_reset (&item);
+      GST_DEBUG_OBJECT (sel, "Activating pad %s:%s",
+          GST_DEBUG_PAD_NAME (active_sinkpad));
+    } else
+      GST_WARNING_OBJECT (sel, "Couldn't find a default sink pad");
+    gst_iterator_free (iter);
+  }
+
+  return active_sinkpad;
+}
+
+/**
+ * @brief request new sink pad
+ */
+static GstPad *
+gst_join_request_new_pad (GstElement * element,
+    GstPadTemplate * templ, const gchar * unused, const GstCaps * caps)
+{
+  GstJoin *sel;
+  gchar *name = NULL;
+  GstPad *sinkpad = NULL;
+
+  g_return_val_if_fail (templ->direction == GST_PAD_SINK, NULL);
+
+  sel = GST_JOIN (element);
+
+  GST_JOIN_LOCK (sel);
+
+  GST_LOG_OBJECT (sel, "Creating new pad sink_%u", sel->padcount);
+  name = g_strdup_printf ("sink_%u", sel->padcount++);
+  sinkpad = g_object_new (GST_TYPE_JOIN_PAD,
+      "name", name, "direction", templ->direction, "template", templ, NULL);
+  g_free (name);
+
+  sel->n_pads++;
+
+  gst_pad_set_event_function (sinkpad, GST_DEBUG_FUNCPTR (gst_join_pad_event));
+  gst_pad_set_query_function (sinkpad, GST_DEBUG_FUNCPTR (gst_join_pad_query));
+  gst_pad_set_chain_function (sinkpad, GST_DEBUG_FUNCPTR (gst_join_pad_chain));
+  gst_pad_set_iterate_internal_links_function (sinkpad,
+      GST_DEBUG_FUNCPTR (gst_join_pad_iterate_linked_pads));
+
+  GST_OBJECT_FLAG_SET (sinkpad, GST_PAD_FLAG_PROXY_CAPS);
+  GST_OBJECT_FLAG_SET (sinkpad, GST_PAD_FLAG_PROXY_ALLOCATION);
+  gst_pad_set_active (sinkpad, TRUE);
+  gst_element_add_pad (GST_ELEMENT (sel), sinkpad);
+  GST_JOIN_UNLOCK (sel);
+
+  return sinkpad;
+}
+
+/**
+ * @brief register this element
+ */
+static gboolean
+plugin_init (GstPlugin * plugin)
+{
+  GST_DEBUG_CATEGORY_INIT (GST_CAT_DEFAULT, "join", 0,
+      "gstreamer join element");
+
+  if (!gst_element_register (plugin, "join", GST_RANK_NONE, GST_TYPE_JOIN)) {
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+#ifndef PACKAGE
+#define PACKAGE "join"
```

Comment:
BTW, what's the hierarchy of plugin and elements? I think `join` is close to an element, not a plugin. (unless this plugin will contain only a single element)
So, we have to think its plugin name, for example
```
plugin: nnstreamer-extra
- elements: join, ...
```

Reply:
Thanks for the suggestion. As we discussed offline, `nnstreamer-misc` seems more suitable as a package name.